### PR TITLE
gh-117398: Isolate _datetime (3.13.0a6 PoC)

### PR DESCRIPTION
**Pull request by neonene**
Adjust the `_datetime` isolation draft (#102995) to main.  This also includes #117413.

Co-authored-by: Erlend E. Aasland <erlend@python.org>

* Issue: gh-117398

### CODE REVIEW

**Comment by neonene** [comment]
@Eclips4 Could you run the refleak bots?  Also, it would be nice if this could fix some sub-interpreter issues you are aware of.

File: Include/datetime.h [hunk 1]
```diff
@@ -194,10 +194,24 @@ typedef struct {
  * */
 #ifndef _PY_DATETIME_IMPL
 /* Define global variable for the C API and a macro for setting it. */
-static PyDateTime_CAPI *PyDateTimeAPI = NULL;
+static PyDateTime_CAPI *_pydatetimeapi_main = NULL;
 
-#define PyDateTime_IMPORT \
-    PyDateTimeAPI = (PyDateTime_CAPI *)PyCapsule_Import(PyDateTime_CAPSULE_NAME, 0)
+static inline void _import_pydatetime(void) {
```
Comment (neonene) [inline Include/datetime.h:199]:
```suggestion
static inline void 
_import_pydatetime(void) {
```

File: Include/datetime.h [hunk 2]
```diff
@@ -194,10 +194,24 @@ typedef struct {
...
+    if (PyInterpreterState_Get() == PyInterpreterState_Main()) {
+        _pydatetimeapi_main = PyCapsule_Import(PyDateTime_CAPSULE_NAME, 0);
+    }
+}
+#define PyDateTime_IMPORT _import_pydatetime()
+
+static inline PyDateTime_CAPI *_get_pydatetime_api(void) {
```
Comment (neonene) [inline Include/datetime.h:206]:
```suggestion
static inline PyDateTime_CAPI *
_get_pydatetime_api(void) {
```

**Comment by erlend-aasland** [comment]
You need to update globals-to-fix.

**Comment by ericsnowcurrently** [comment]
Thanks @neonene and @erlend-aasland for this PR.  It was a really helpful reference for the changes I made recently.

**Pull request closed without merging**